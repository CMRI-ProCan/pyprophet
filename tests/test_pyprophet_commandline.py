import os
import subprocess
import shutil
import sys

import pytest

__here__ = os.path.dirname(os.path.abspath(__file__))


def _record(stdout, regtest):
    for line in stdout.split("\n"):
        if line.startswith("INFO"):
            # remove data which is not constant over runs as time or process id:
            line = line.split(":", 4)[-1]
        if "NEEDED" in line:
            continue
        if "time needed" in line:
            continue
        print >> regtest, line


def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError, e:
        print >> sys.stderr, e.output
        raise
    return stdout


def _run_pyprophet_to_learn_model(regtest, temp_folder, dump_result_files=False, with_probatilities=False):
    os.chdir(temp_folder)
    data_path = os.path.join(__here__, "test_data.txt")
    shutil.copy(data_path, temp_folder)
    if with_probatilities:
        stdout = _run_cmdline("pyprophet test_data.txt --random_seed=42 --compute.probabilities")
    else:
        stdout = _run_cmdline("pyprophet test_data.txt --random_seed=42")
    for f in ["test_data_summary_stat.csv", "test_data_full_stat.csv", "test_data_report.pdf",
              "test_data_scorer.bin",
              "test_data_cutoffs.txt",
              "test_data_svalues.txt",
              "test_data_qvalues.txt",
              "test_data_with_dscore.csv",
              "test_data_with_dscore_filtered.csv",
              "test_data_dscores_top_target_peaks.txt",
              "test_data_dscores_top_decoy_peaks.txt",
              ]:
        full_path = os.path.join(temp_folder, f)
        assert os.path.exists(full_path)

    if not dump_result_files:
        return stdout

    for f in ["test_data_summary_stat.csv", "test_data_full_stat.csv",
              "test_data_cutoffs.txt",
              "test_data_svalues.txt",
              "test_data_qvalues.txt",
              "test_data_dscores_top_target_peaks.txt",
              "test_data_dscores_top_decoy_peaks.txt",
              "test_data_with_dscore.csv",
              "test_data_with_dscore_filtered.csv",
              ]:
        full_path = os.path.join(temp_folder, f)
        dump(full_path, regtest)

    full_path = os.path.join(temp_folder, "test_data_scorer.bin")
    dump_digest(regtest, full_path, "test_data_scorer.bin")

    return stdout

def dump(full_path, regtest):
    lines = open(full_path, "r").readlines()
    f = os.path.basename(full_path)
    print >> regtest
    print >> regtest, f, "contains", len(lines), "lines"
    print >> regtest
    if len(lines) > 10:
        print >> regtest, "top 5 lines of", f
        for line in lines[:5]:
            if len(line) > 80:
                line = line[:40] + " ... " + line[-45:]
            print >> regtest, line.rstrip()
        print >> regtest
        print >> regtest, "last 5 lines of", f
        for line in lines[-5:]:
            if len(line) > 80:
                line = line[:40] + " ... " + line[-45:]
            print >> regtest, line.rstrip()
    else:
        for line in lines:
            if len(line) > 80:
                line = line[:40] + " ... " + line[-45:]
            print >> regtest, line.rstrip()



def test_0(tmpdir, regtest):
    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True, False)
    _record(stdout, regtest)


def test_1(tmpdir, regtest):
    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True, True)
    _record(stdout, regtest)


def test_apply_classifier(tmpdir, regtest):

    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True)

    _record(stdout, regtest)

    full_path = os.path.join(tmpdir.strpath, "test_data_with_dscore.csv")
    dump(full_path, regtest)

    full_path = os.path.join(tmpdir.strpath, "test_data_with_dscore_filtered.csv")
    dump(full_path, regtest)

    # collect m score stats
    m_score_stat, = [l for l in stdout.split("\n") if "mean m_score" in l]

    # split away log time etc:
    __, __, interesting_m_score_output = m_score_stat.partition("mean m_score")

    # collect s value stats
    s_value_stat, = [l for l in stdout.split("\n") if "mean s_value" in l]

    # split away log time etc:
    __, __, interesting_s_value_output = s_value_stat.partition("mean s_value")

    stdout = _run_cmdline("pyprophet test_data.txt --apply_scorer=test_data_scorer.bin "
                          "--target.overwrite --random_seed=42")

    _record(stdout, regtest)

    # collect m score stats
    m_score_stat, = [l for l in stdout.split("\n") if "mean m_score" in l]

    # split away log time etc:
    __, __, interesting_m_score_output2 = m_score_stat.partition("mean m_score")

    # collect s value stats
    s_value_stat, = [l for l in stdout.split("\n") if "mean s_value" in l]

    # split away log time etc:
    __, __, interesting_s_value_output2 = s_value_stat.partition("mean s_value")

    # check if output is consistent over runs:
    assert interesting_m_score_output == interesting_m_score_output2
    assert interesting_s_value_output == interesting_s_value_output2


def test_apply_weights(tmpdir, regtest):

    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir.strpath, True)
    _record(stdout, regtest)

    stdout = _run_cmdline("pyprophet test_data.txt --apply_weights=test_data_weights.txt "
                          "--target.overwrite --random_seed=42")

    _record(stdout, regtest)

    for name in ["test_data_summary_stat.csv",
                 "test_data_full_stat.csv",
                 "test_data_report.pdf",
                 "test_data_cutoffs.txt",
                 "test_data_svalues.txt",
                 "test_data_qvalues.txt",
                 "test_data_dscores_top_target_peaks.txt",
                 "test_data_dscores_top_decoy_peaks.txt",
                 "test_data_with_dscore.csv",
                 "test_data_with_dscore_filtered.csv", ]:

        full_path = os.path.join(tmpdir.strpath, name)
        dump(full_path, regtest)

    full_path = os.path.join(tmpdir.strpath, "test_data_scorer.bin")
    dump_digest(regtest, full_path, "test_data_scorer.bin")


def dump_digest(regtest, full_path, name):
    import hashlib
    h = hashlib.sha1()
    h.update(open(full_path, "rb").read())
    print >> regtest
    print >> regtest, "digest", name, ":", h.hexdigest()



def test_multiple_input_files(tmpdir, regtest):
    os.chdir(tmpdir.strpath)

    data_path = os.path.join(__here__, "test_data_3.txt")
    shutil.copy(data_path, tmpdir.strpath)
    data_path = os.path.join(__here__, "test_data_2.txt")
    shutil.copy(data_path, tmpdir.strpath)

    stdout = _run_cmdline("pyprophet test_data_2.txt test_data_3.txt --random_seed=42 --target.overwrite")
    _record(stdout, regtest)


def test_multiple_input_files_with_merge(tmpdir, regtest):
    os.chdir(tmpdir.strpath)

    data_path = os.path.join(__here__, "test_data_3.txt")
    shutil.copy(data_path, tmpdir.strpath)
    data_path = os.path.join(__here__, "test_data_2.txt")
    shutil.copy(data_path, tmpdir.strpath)

    stdout = _run_cmdline("pyprophet test_data_2.txt test_data_3.txt --random_seed=42 --target.overwrite "
                          "--multiple_files.merge_results --compute.probabilities")

    _record(stdout, regtest)


def test_out_of_core_multi_input_files(tmpdir, regtest):

    def setup(subfolder):
        f = tmpdir.join(subfolder).strpath
        os.makedirs(f)
        os.chdir(f)
        data_path = os.path.join(__here__, "test_data_3.txt")
        shutil.copy(data_path, f)
        data_path = os.path.join(__here__, "test_data_2.txt")
        shutil.copy(data_path, f)
        return f

    f1 = setup("out_of_core_1.0")
    stdout = _run_cmdline("pyprophet test_data_2.txt test_data_3.txt "
                          "--out_of_core "
                          "--out_of_core.sampling_rate=1.0 --random_seed=42")
    _record(stdout, regtest)

    f1merge = setup("out_of_core_1.0_merged")
    stdout = _run_cmdline("pyprophet test_data_2.txt test_data_3.txt "
                          "--out_of_core "
                          "--out_of_core.sampling_rate=1.0 --random_seed=42 "
                          "--multiple_files.merge_results")

    _record(stdout, regtest)

    f2 = setup("in_core")
    stdout = _run_cmdline("pyprophet test_data_2.txt test_data_3.txt --random_seed=42")
    _record(stdout, regtest)

    f2merge = setup("in_core_merged")
    stdout = _run_cmdline("pyprophet test_data_2.txt test_data_3.txt --random_seed=42 "
                          "--multiple_files.merge_results")
    _record(stdout, regtest)

    f3 = setup("out_of_core_0.1")
    stdout = _run_cmdline("pyprophet test_data_2.txt test_data_3.txt "
                          "--out_of_core --out_of_core.sampling_rate=0.9999999 --random_seed=42")
    _record(stdout, regtest)

    print >> regtest
    print >> regtest, "the next three should be identical:"
    full_path = os.path.join(f1, "test_data__summary_stat.csv")
    dump(full_path, regtest)
    full_path = os.path.join(f1merge, "test_data__summary_stat.csv")
    dump(full_path, regtest)
    full_path = os.path.join(f2, "test_data__summary_stat.csv")
    dump(full_path, regtest)
    print >> regtest
    print >> regtest, "this might be different:"
    full_path = os.path.join(f3, "test_data__summary_stat.csv")
    dump(full_path, regtest)

    compare_folders(f1, f2)
    compare_folders(f1merge, f2merge)


def compare_folders(f1, f2):

    f1files = sorted([os.path.basename(fi) for fi in os.listdir(f1)])
    f2files = sorted([os.path.basename(fi) for fi in os.listdir(f2)])

    assert f1files == f2files

    for name in f1files:
        lines1 = open(os.path.join(f1, name), "r").readlines()
        lines2 = open(os.path.join(f1, name), "r").readlines()
        n1 = len(lines1)
        n2 = len(lines2)
        assert len(lines1) == len(lines2), (name, n1, n2)
        for i, (l1, l2) in enumerate(zip(lines1, lines2)):
            assert l1 == l2, (name, i, l1, l2)
