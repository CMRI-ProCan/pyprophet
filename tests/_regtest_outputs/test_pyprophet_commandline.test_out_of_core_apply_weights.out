pyprophet test_data_2.txt test_data_3.txt --out_of_core --out_of_core.sampling_rate=1.0 --random_seed=42 --multiple_files.merge_results
 config settings:
     apply_scorer: None
     apply_weights: None
     compute.probabilities: False
     d_score.cutoff: -1000.0
     delim.in:
     delim.out:
     export.mayu: False
     final_statistics.emp_p: False
     final_statistics.lambda: 0.4
     final_statistics.lfdr_eps: 1e-08
     final_statistics.lfdr_monotone: True
     final_statistics.lfdr_transf: probit
     final_statistics.lfdr_trunc: True
     final_statistics.pfdr: False
     final_statistics.pi0_method: smoother
     final_statistics.pi0_smooth_df: 3
     final_statistics.pi0_smooth_log_pi0: False
     group_id: transition_group_id
     ignore.invalid_score_columns: 0
     is_test: 0
     multiple_files.merge_results: True
     num_processes: 1
     out_of_core: 1
     out_of_core.sampling_rate: 1.0
     random_seed: 42
     semi_supervised_learner.initial_fdr: 0.15
     semi_supervised_learner.iteration_fdr: 0.02
     semi_supervised_learner.num_iter: 5
     semi_supervised_learner.stat_best: 0
     semi_supervised_learner.use_best: 0
     target.compress_results: 0
     target.dir: None
     target.overwrite: 0
     target.prefix: None
     xeval.fraction: 0.5
     xeval.num_iter: 5
 learn and apply classifier out of core
 data set contains 387 decoy and 387 target transition groups
 data set contains 387 decoy and 387 target transition groups
 summary input file:
    18330 lines
    774 transition groups
    17 scores including main score
 learn and apply scorer
 start 5 cross evals using 1 processes
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 finished cross evals

 Estimate local false discovery rates (lFDR) / posterior error probabilities (PEP)
 calculated scoring and statistics
 processing input data finished
 data set contains 387 decoy and 387 target transition groups
 mean m_score = 6.130687e-02, std_dev m_score = 1.658475e-02
 mean s_value = 9.800913e-01, std_dev s_value = 1.125709e-01
 data set contains 387 decoy and 387 target transition groups
 mean m_score = 6.130687e-02, std_dev m_score = 1.658475e-02
 mean s_value = 9.800913e-01, std_dev s_value = 1.125709e-01

used parameters:

    apply_scorer                               : None
    apply_weights                              : None
    compute.probabilities                      : False
    d_score.cutoff                             : -1000.0
    delim.in                                   : '\t'
    delim.out                                  : '\t'
    export.mayu                                : False
    final_statistics.emp_p                     : False
    final_statistics.lambda                    : 0.4
    final_statistics.lfdr_eps                  : 1e-08
    final_statistics.lfdr_monotone             : True
    final_statistics.lfdr_transf               : 'probit'
    final_statistics.lfdr_trunc                : True
    final_statistics.pfdr                      : False
    final_statistics.pi0_method                : 'smoother'
    final_statistics.pi0_smooth_df             : 3
    final_statistics.pi0_smooth_log_pi0        : False
    group_id                                   : 'transition_group_id'
    ignore.invalid_score_columns               : 0
    is_test                                    : 0
    multiple_files.merge_results               : True
    num_processes                              : 1
    out_of_core                                : 1
    out_of_core.sampling_rate                  : 1.0
    random_seed                                : 42
    semi_supervised_learner.initial_fdr        : 0.15
    semi_supervised_learner.iteration_fdr      : 0.02
    semi_supervised_learner.num_iter           : 5
    semi_supervised_learner.stat_best          : 0
    semi_supervised_learner.use_best           : 0
    target.compress_results                    : 0
    target.dir                                 : None
    target.overwrite                           : 0
    target.prefix                              : None
    xeval.fraction                             : 0.5
    xeval.num_iter                             : 5


==================================================================================================

   qvalue   svalue        pvalue           pep     TP    FP        TN         FN           FDR  \
0    0.00  0.00223  7.172041e-14  7.205829e-09    2.0   0.0  54.80829  717.19171  2.450447e-12
1    0.01  1.00000  1.178847e-01  5.291491e-01  720.0   6.0  48.80829   -0.80829  8.901058e-03
2    0.02  1.00000  2.421850e-01  1.000000e+00  721.0  13.0  41.80829   -1.80829  1.808667e-02
3    0.05  1.00000  6.864012e-01  1.000000e+00  718.0  38.0  16.80829    1.19171  4.973922e-02
4    0.10  1.00000  9.622872e-01  1.000000e+00  721.0  53.0   1.80829   -1.80829  6.814124e-02
5    0.20      NaN           NaN           NaN    NaN   NaN       NaN        NaN           NaN
6    0.30      NaN           NaN           NaN    NaN   NaN       NaN        NaN           NaN
7    0.40      NaN           NaN           NaN    NaN   NaN       NaN        NaN           NaN
8    0.50      NaN           NaN           NaN    NaN   NaN       NaN        NaN           NaN

        FNR      sens    cutoff
0  0.929041  0.002781  7.229383
1  0.000000  1.000000  1.680954
2  0.000000  1.000000  1.045573
3  0.025943  0.998343 -0.363325
4  0.000000  1.000000 -1.586764
5       NaN       NaN       NaN
6       NaN       NaN       NaN
7       NaN       NaN       NaN
8       NaN       NaN       NaN

==================================================================================================

WRITTEN:  test_data__summary_stat.csv
WRITTEN:  test_data__full_stat.csv
WRITTEN:  test_data__with_dscore.csv
WRITTEN:  test_data__with_dscore_filtered.csv
WRITTEN:  test_data__report.pdf
WRITTEN:  test_data__cutoffs.txt
WRITTEN:  test_data__svalues.txt
WRITTEN:  test_data__qvalues.txt
WRITTEN:  test_data__dscores_top_target_peaks.txt
WRITTEN:  test_data__dscores_top_decoy_peaks.txt
WRITTEN:  test_data__scorer.bin
WRITTEN:  test_data__weights.txt


pyprophet test_data_2.txt test_data_3.txt --out_of_core --apply_weights=test_data__weights.txt --out_of_core.sampling_rate=1.0 --random_seed=42 --target.overwrite --multiple_files.merge_results
 config settings:
     apply_scorer: None
     apply_weights: test_data__weights.txt
     compute.probabilities: False
     d_score.cutoff: -1000.0
     delim.in:
     delim.out:
     export.mayu: False
     final_statistics.emp_p: False
     final_statistics.lambda: 0.4
     final_statistics.lfdr_eps: 1e-08
     final_statistics.lfdr_monotone: True
     final_statistics.lfdr_transf: probit
     final_statistics.lfdr_trunc: True
     final_statistics.pfdr: False
     final_statistics.pi0_method: smoother
     final_statistics.pi0_smooth_df: 3
     final_statistics.pi0_smooth_log_pi0: False
     group_id: transition_group_id
     ignore.invalid_score_columns: 0
     is_test: 0
     multiple_files.merge_results: True
     num_processes: 1
     out_of_core: 1
     out_of_core.sampling_rate: 1.0
     random_seed: 42
     semi_supervised_learner.initial_fdr: 0.15
     semi_supervised_learner.iteration_fdr: 0.02
     semi_supervised_learner.num_iter: 5
     semi_supervised_learner.stat_best: 0
     semi_supervised_learner.use_best: 0
     target.compress_results: 0
     target.dir: None
     target.overwrite: 1
     target.prefix: None
     xeval.fraction: 0.5
     xeval.num_iter: 5
 apply weights out of core
 data set contains 387 decoy and 387 target transition groups
 data set contains 387 decoy and 387 target transition groups
 summary input file:
    18330 lines
    774 transition groups
    17 scores including main score
 start application of pretrained weights
 finished pretrained scoring
 Estimate local false discovery rates (lFDR) / posterior error probabilities (PEP)
 calculated scoring and statistics
 processing input data finished
 data set contains 387 decoy and 387 target transition groups
 mean m_score = 5.802908e-02, std_dev m_score = 1.645686e-02
 mean s_value = 9.800469e-01, std_dev s_value = 1.126820e-01
 data set contains 387 decoy and 387 target transition groups
 mean m_score = 5.802908e-02, std_dev m_score = 1.645686e-02
 mean s_value = 9.800469e-01, std_dev s_value = 1.126820e-01

used parameters:

    apply_scorer                               : None
    apply_weights                              : 'test_data__weights.txt'
    compute.probabilities                      : False
    d_score.cutoff                             : -1000.0
    delim.in                                   : '\t'
    delim.out                                  : '\t'
    export.mayu                                : False
    final_statistics.emp_p                     : False
    final_statistics.lambda                    : 0.4
    final_statistics.lfdr_eps                  : 1e-08
    final_statistics.lfdr_monotone             : True
    final_statistics.lfdr_transf               : 'probit'
    final_statistics.lfdr_trunc                : True
    final_statistics.pfdr                      : False
    final_statistics.pi0_method                : 'smoother'
    final_statistics.pi0_smooth_df             : 3
    final_statistics.pi0_smooth_log_pi0        : False
    group_id                                   : 'transition_group_id'
    ignore.invalid_score_columns               : 0
    is_test                                    : 0
    multiple_files.merge_results               : True
    num_processes                              : 1
    out_of_core                                : 1
    out_of_core.sampling_rate                  : 1.0
    random_seed                                : 42
    semi_supervised_learner.initial_fdr        : 0.15
    semi_supervised_learner.iteration_fdr      : 0.02
    semi_supervised_learner.num_iter           : 5
    semi_supervised_learner.stat_best          : 0
    semi_supervised_learner.use_best           : 0
    target.compress_results                    : 0
    target.dir                                 : None
    target.overwrite                           : 1
    target.prefix                              : None
    xeval.fraction                             : 0.5
    xeval.num_iter                             : 5


==================================================================================================

   qvalue    svalue        pvalue           pep     TP    FP         TN          FN           FDR  \
0    0.00  0.011101  3.013145e-13  6.821255e-09    8.0   0.0  53.333333  712.666667  2.008764e-12
1    0.01  1.000000  1.478791e-01  8.660165e-01  726.0   8.0  45.333333   -5.333333  1.074508e-02
2    0.02  1.000000  2.900465e-01  1.000000e+00  725.0  15.0  38.333333   -4.333333  2.090426e-02
3    0.05  1.000000  6.793253e-01  1.000000e+00  722.0  36.0  17.333333   -1.333333  4.779774e-02
4    0.10  1.000000  9.437169e-01  1.000000e+00  724.0  50.0   3.333333   -3.333333  6.502787e-02
5    0.20       NaN           NaN           NaN    NaN   NaN        NaN         NaN           NaN
6    0.30       NaN           NaN           NaN    NaN   NaN        NaN         NaN           NaN
7    0.40       NaN           NaN           NaN    NaN   NaN        NaN         NaN           NaN
8    0.50       NaN           NaN           NaN    NaN   NaN        NaN         NaN           NaN

        FNR      sens    cutoff
0  0.930374  0.011101  7.199885
1  0.000000  1.000000  1.045573
2  0.000000  1.000000  0.553249
3  0.000000  1.000000 -0.465813
4  0.000000  1.000000 -1.586764
5       NaN       NaN       NaN
6       NaN       NaN       NaN
7       NaN       NaN       NaN
8       NaN       NaN       NaN

==================================================================================================

WRITTEN:  test_data__summary_stat.csv
WRITTEN:  test_data__full_stat.csv
WRITTEN:  test_data__with_dscore.csv
WRITTEN:  test_data__with_dscore_filtered.csv
WRITTEN:  test_data__report.pdf
WRITTEN:  test_data__cutoffs.txt
WRITTEN:  test_data__svalues.txt
WRITTEN:  test_data__qvalues.txt
WRITTEN:  test_data__dscores_top_target_peaks.txt
WRITTEN:  test_data__dscores_top_decoy_peaks.txt
WRITTEN:  test_data__scorer.bin


